---
title: "Molecular adaption of cancer associated genes"
author: "Alberto Vicens Sanchez"
date: "February 19, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, eval = FALSE, warning = F,cache = T, comment = F,  cache.path="/home/user/phylogenomics_lab/cancer_genes_selection/cache/",autodep=T, fig.path="/home/user/Dropbox/phylogenomics_lab_dbx/cancer_genes_selection/figures/")
```

# Introduction 
In this study, we will evaluate the selective pressures driving the evolution of human cancer associated genes, spanning both oncogenes and tumor supressor genes (TSG), in a phylogeny of mammals. 

The objectives of this study are:

* Assess selective pressures on cancer driver genes in mammals.
* Compare selective regimes between different functional categories of cancer genes.
* Compare germline evolution against somatic variation.
* Identify episodes of positive selection on cancer genes across mamalian species.

# Species
For this study, we compared 32 mammalian species with well annotated genomes. Information about the species see the table:
https://drive.google.com/file/d/1GpDmoZHW6ZdfqBQ8Bn0OPM53OfESNrkt/view?usp=sharing

## Cancer genes
I retrieved cancer genes from the COSMIC database. A list of 574 genes classified as Tier 1 (i.e. those genes with a documented activity in cancer) was downloaded.
https://drive.google.com/file/d/1GmpOnIMWBBgYBJZaUulY9MUOIgEO9g8I/view?usp=sharing

I created a compacted data frame from COSMIC table including following information: 

* Somatic
* Germline
* Tumor type
* Tissue type
* Molecular genetics
* Role in cancer
* Mutation type
```{bash compact_database}
cat cosmic_cancer_genes.tsv | cut -f1,2,8,9,10,13,14,15,16 | head -n -4 > cosmic_cancer_genes_compacted.tsv 
```

## Retrieving human cancer genes
I retrieved information for human genes using Ensembl BioMArt (see *ensemble_gene_id.tsv* file). 
After applying several supporting factors for gene annotation and transcript support level, I finally obtained 535 supported genes.
```{bash, filter_genes}
tail -n +2 ensembl_gene_list.tsv| grep "\<tsl1\>" | grep 'GENCODE'| sort -k4,4 -k6nr,6 | sort -u -k 4,4 > ensembl_gene_uniq.tsv
cut -f4 ensembl_gene_uniq.tsv > genenames.txt
```

The following genes did not pass the supporting filters and were discarded from the analysis:
*ACKR3,AFF3,CEBPA,FANCF,FAT1,FOXL2,HIST1H3B,HIST1H4I,HLA-A,HOXD11,HOXD13,IRS4,JAK1,JUN,KMT2D,MTCP1,NUTM2B,PIK3CA,POT1,SMAD4,SOX2,TAL2*

To organize the data per gene, I create a working folder for each gene with the *create_gene_folders.sh* script.
```{bash create_gene_folders}
#!/bin/env bash

WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
DATADIR="${WORKDIR}/data"
GENEDIR="${WORKDIR}/genes"

while read gene; do
mkdir ${GENEDIR}/$gene
done < ${DATADIR}/genenames.txt
```

#Retrieving Orthologs
##Get human peptide sequences
I extract the Ensembl protein ID from the human protein infromation table
```{bash}
cut -f2 ensembl_gene_uniq.tsv > protein_ensembl_id.txt
```
I retrieve peptide sequences of human proteins with BioMart (see *human_pep_seqs.fasta*), then I move the retrieved human protein sequences to the respective gene folder.
```{bash}
#!/bin/bash
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
DATADIR="${WORKDIR}/data"
GENEDIR="${WORKDIR}/genes"
GENENAMES=`ls ${GENEDIR}`
HUMANSEQS="${DATADIR}/human_pep_seqs_one_line.fasta"

while read -r gene; do
SEQFILE=`grep "|${gene}$" ${HUMANSEQS} | sed -e 's/^>//' -e 's/|/_/' -e 's/.*$/&.fas/'`
grep -A1 "|${gene}$" ${HUMANSEQS} > ${GENEDIR}/${gene}/${SEQFILE}
done <<< "${GENENAMES}"
```

#Get orthologs for human sequences
I use the BiomaRt library implemented in R to get orthologues for all genes and from all species. 
```{r get_orthologues, eval=F, dpi=100, fig.width=8, fig.height=6, fig.cap= "Figure 1. Heatmap representing the presence/absence pattern of 535 cancer genes across 32 mammalian genomes."}
library(biomaRt)

#Load database
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")

#Obtain list of orthologues
spfile<-read.table("/home/user/Dropbox/phylogenomics_lab_dbx/cancer_genes_selection/data/species_ensembl_code.txt",header = F)
splist<-paste(spfile$V1,"_homolog_ensembl_peptide",sep="")
splist=splist[-11] #Discard human

#Obtain list of human peptide ID
pepfile<-read.table("/home/user/Dropbox/phylogenomics_lab_dbx/cancer_genes_selection/data/protein_ensembl_id.txt",header = F)
peplist=pepfile$V1

#Set attributes
attributes = c("external_gene_name","ensembl_peptide_id")
orthologues = getBM(attributes,filters="ensembl_peptide_id",values=peplist,TRUE, mart = human, uniqueRows=TRUE)
colnames(orthologues)=c("Gene","hsapiens")

for (i in 1:length(splist)) {
ortho=getBM(c("external_gene_name",splist[i]),filters="ensembl_peptide_id",values=peplist, mart = human, uniqueRows=TRUE)
ortho=ortho[!duplicated(ortho$external_gene_name),2]
orthologues=cbind(orthologues,ortho)
spname<-sapply(strsplit(splist[i],"_"), "[",1)
colnames(orthologues)[ncol(orthologues)]=spname
}

#Costruct a binary matrix and plot a heatmap
bintable<-data.frame(genes=as.character(orthologues[,1]))
rownames(bintable)= bintable$genes
bintable[,1]=NULL
orthomat=as.matrix(orthologues[,-1])
binmat=matrix(,nrow=nrow(orthomat),ncol=ncol(orthomat),dimnames=list(orthologues$Gene,colnames(orthomat)))

for (i in 1:length(orthomat)) {
    if (orthomat[i] == "") {
      binmat[i] = 0
    }
    
    else {binmat[i]=1}
  }

heatmap(binmat,scale = "none",cexRow = 0.2, cexCol = 0.7)
```

```{r, create_genetable}
#Create gene table and add number of retrieved sequences per gene
nseqs<-apply(binmat,1,function(x) length(x[x==1]))
genetable<-data.frame(genes=names(nseqs),seqs=nseqs,row.names = NULL)
save(genetab)

write.table(genetable,file = "Dropbox/phylogenomics_lab_dbx/cancer_genes_selection/tables/cancer_genes_table.tsv",row.names = F,sep="\t",col.names = T, quote=F)
```
In this step, I discarded those genes with < 20 orthologues sequences

I then download ortholog coding sequences from BioMart
```{bash, download_ortholog_cds}
#!/bin/bash

#SBATCH -N 1
#SBATCH -n 1
#SBATCH --cpus-per-task 4
#SBATCH -t 00:30:00
#SBATCH --mem 6G

#Setting directories and files
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
DATADIR="${WORKDIR}/data"
GENEDIR="${WORKDIR}/genes"
DBDIR="${WORKDIR}/db/pep"
SCRIPTDIR="${WORKDIR}/scripts"
SPNAMES=${DATADIR}/species_codes.txt

GENE=`ls ${GENEDIR} | sed "${SLURM_ARRAY_TASK_ID}q;d"`
echo "Creating directory for coding sequences of ${GENE}"
mkdir ${GENEDIR}/${GENE}/seqs
mkdir ${GENEDIR}/${GENE}/seqs/cds
protlist="${GENEDIR}/${GENE}/*_orthologs.txt"

#Load PERL module
module load perl/5.24.0
#Set path to Biomart modules
export PERL5LIB=$PERL5LIB:/home/uvi/be/avs/tools/biomart-perl/lib/:/home/uvi/be/avs/tools/biomart-perl/perl_modules/

while read -r id; do
spcode=`echo $id | cut -c -6`
dataset=`grep $spcode ${SPNAMES} | cut -d' ' -f2`
echo "Downloading ortholog coding sequences of ${GENE} from ${dataset}"
perl ${SCRIPTDIR}/download_cds_from_protid.pl $id $dataset >> ${GENEDIR}/${GENE}/seqs/cds/"${GENE}_cds.fas"
done < ${protlist}
```

#Multiple sequence alignment (MSA)
##Codon based alignment
I perform MSA of coding nucleotide sequences using MACSE (Ranwez et al. 2011)
```{bash, multiple_alignment}
#!/bin/bash

#SBATCH -N 1
#SBATCH -n 1
#SBATCH --cpus-per-task 8
#SBATCH -t 00:45:00
#SBATCH --mem 6G

#Setting directories and files
TOOLSDIR="/home/uvi/be/avs/tools"
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
GENEDIR="${WORKDIR}/genes"

GENE=`ls ${GENEDIR} | sed "${SLURM_ARRAY_TASK_ID}q;d"`
echo "Creating directory for alignment of ${GENE}"
mkdir ${GENEDIR}/${GENE}/align
seqfile="${GENEDIR}/${GENE}/seqs/${GENE}_align.fas"

module load jdk/1.8.0

echo "Running codon alignment of ${GENE} sequences"
java -jar ${TOOLSDIR}/macse_v0.9b1.jar -i ${seqfile} -g -7 -x -1 -f -30 -d 1 -s -100 -o ${GENE}
```
In this step, I discarded 3 genes due to an excessive computing time for codon alignment: KMT2C, RNF213, ZFHX3

##Curating MSA
The MSA curation had the following steps:

* Clean up secondary annotations of fasta headers (just retain Ensembl gene ids)
* Discarding duplicate sequences
* Replacing "!" characters (inserted by Macse to corect frameshits) for "N"

```{bash, curate_msa}
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
GENEDIR="${WORKDIR}/genes"
GENENAMES=`ls ${GENEDIR}`

while read -r gene; do
echo "curating alignment for ${gene}"
cd ${GENEDIR}/${gene}/align/
awk '/^>/{f=!d[$1];d[$1]=1}f' ${gene}_align_DNA.fasta | sed 's/!/N/g' | cut -d'|' -f1 > ${gene}_align_DNA_curated.fasta
done <<< "${GENENAMES}"
```
Once MSA were curated, I proceed to trim the MSAs using TrimAl (Capella-Gutierrez et al. 2009) with following parameters: 
  *Delete columns with gaps in more than 60% of the sequences.
  *Delete columns with a similarity score lower than 0.1
  *Remove sequences not covering at least the 60% of residues that achieve an overlap, with the rest of the sequences, of 0.75.
  
```{bash trim_msa}
#Setting directories and files
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
GENEDIR="${WORKDIR}/genes"
GENENAMES=`ls ${GENEDIR}`

while read -r gene; do
echo "trimming alignment for ${gene}"
trimal -in ${GENEDIR}/${gene}/align/${gene}_align_DNA_curated.fasta -out ${GENEDIR}/${gene}/align/${gene}_align_DNA_trimmed.fasta -gt 0.4 -st 0.1 -resoverlap 0.8 -seqoverlap 70
done <<< "${GENENAMES}"
```

#Discarded genes
After trimming, the following genes were discarded due to their MSA did not pass TrimAl filters:
*ARID1B, LRP1B, POLQ, PTCH1, PTPRK, SMARCA4, TRRAP, TSC2, UBR5, CACNA1D, AFF1, CNTRL, ATP1A1, BRCA1, CUX1, EZR, FBXO11, FUBP1, GAS7, GNAS, MKL1, NRG1, NTRK3, NUP214, PLCG1, RAP1GDS1, TAL1, MAP3k13, PPARG, RUNX1T1,AMER1,ASXL1,CIC,MAP2K4,MSI2,MUC1,NUTM1,OLIG2,QKI,SFPQ,SSX1,SSX2,SSX4,TFEB*

The following genes were discarded for having alignments with <= 10 taxa:
*MTOR MUTYH RARA TNFRSF14 MITF TCF7L2*

#Tree reconstruction
I reconstruct trees for each gene using the program *RAxML-ng*. The analysis included ML tree search + non-parametric bootstrap, with these parameters:
  * 10 randomized parsimony starting trees,
  * General Time Reversible substitution model with discrete GAMMA model of rate heterogeneity with 4 categories (GTR+G)
  * 100 bootstrap replicates
  
```{bash, run_raxml_array}
#!/bin/bash

#SBATCH -N 1
#SBATCH -n 1
#SBATCH --cpus-per-task 8
#SBATCH -t 05:00:00
#SBATCH --mem 6G

#Setting directories and files
TOOLSDIR="/home/uvi/be/avs/tools"
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
GENEDIR="${WORKDIR}/genes"

gene=`ls ${GENEDIR} | sed "${SLURM_ARRAY_TASK_ID}q;d"`
echo "Creating directory for tree of ${GENE}"
mkdir ${GENEDIR}/${gene}/tree

module load gcc/5.3.0 openmpi/1.10.2 raxml-ng/0.5.1b

echo "Building tree for ${gene}"
raxml-ng-mpi --all --msa ${GENEDIR}/${gene}/align/${gene}_align_DNA_trimmed.fas --model GTR+G -tree pars{10} --bs-trees 100 --prefix ${GENEDIR}/${gene}/tree/${gene}
```

##Comparing gene trees with species tree
```{bash, compare_trees}
cd $LUSTRE/cancer_genes_selection
while read -r gene; do
echo /home/uvi/be/avs/lustre/cancer_genes_selection/genes/${gene}/tree/${gene}.raxml.bestTree; done < data/genelist > treelist

module load miniconda
source activate /home/uvi/be/avs/tools

cat treelist | ete3 compare -r species_tree.nwk --src_attr_parser '(^.{6})' --unrooted > tree_distances.txt
```

```{r tree_distances, eval=TRUE, fig.cap="Figure . Frequency histogram representing distances of cancer gene trees with respect to the mammalian species tree"}

process.dist.table <- function(dt,gl) {
  
  dt<-dt[,-grep("X.",colnames(dt))]
  dt<-dt[-1,-c(1,2,10)]
  
  dt=lapply(dt,function (x) { as.numeric(levels(x)[x])})
  dt=as.data.frame(dt)
  dt=cbind(gl,dt)
  }

dt<-read.table("/home/user/Dropbox/phylogenomics_lab_dbx/cancer_genes_selection/tree_analysis/tree_distances466.txt",header=T)
genelist<-read.table("/home/user/Dropbox/phylogenomics_lab_dbx/cancer_genes_selection/data/genelist466.txt", header=T)
dt<-process.dist.table(dt,genelist)

hist(dt$nRF,xlab="Normalized Robinson-Foulds distance",sub="Mammalian tree, 35 taxa",col="lightblue",main=NULL) 
```
The following genes were discarded for having a distance > 0.6 respect a robust species tree of 19 taxa:
CBFB CD74 H3F3B KRAS LMO2 BAX BTG1 CHCHD7 ETV1 FCGR2B FGFR2 LEF1 LMO1 MAX MLF1 PRRX1 RAC1 SOCS1 TPM4 ZRSR2
BCL11A DNM2 FBXW7 H3F3B KRAS LMO2 PAX7 RSPO3 SMARCE1 TBL1XR1

I also built a species supertree given the set of gene trees using Astral-III (Zhang et al. 2017).
```{bash build_supertree}
#!/bin/bash

#Setting directories and files
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
GENEDIR="${WORKDIR}/genes"

TREEFILE="genetrees.nwk"
GENELIST=`ls ${GENEDIR}`

while read -r gene; do
GENETREE="${GENEDIR}/${gene}/tree/${gene}_trimmed.raxml.bestTree"
sed -e 's/ENSG/ENSHSAG/g' -e 's/ENS\(.\{3\}\)\(.\{12\}\)/\1/g' ${GENETREE} > ${WORKDIR}/tmptree
cat ${WORKDIR}/${TREEFILE} ${WORKDIR}/tmptree > tmpfile
mv tmpfile ${WORKDIR}/${TREEFILE}
done <<< "$GENELIST"
```
```{r plot_supertree,fig.align="center",dpi=100,fig.height=8,fig.width=6,fig.cap="Figure. Species tree consturcted given the set of 466 gene trees. Node labels indicate branch support values measure the support for a quadripartition (the four clusters around a branch)"}
library(ape)
suptree<-read.tree("/home/user/Dropbox/phylogenomics_lab_dbx/cancer_genes_selection/data/cancergenes_astral.tre")
plot(suptree); nodelabels(suptree$node.label)
```

#Evolutionary analysis
I run the Codeml program implemented in PAML (Yang 2007) using the ETE3 evol package (Huerta-Cepas et al. 2006). The set parameters in common for all models were:

  * Codon frequency: codon table
  * Initial omega: 0.7
  * Number of gamma categories: 4

```{bash, run_paml}
#!/bin/bash

#SBATCH -N 1
#SBATCH -n 1
#SBATCH -c 2
#SBATCH --cpus-per-task 7
#SBATCH -t 10:00:00
#SBATCH --mem 6G

#Setting directories and files
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
GENEDIR="${WORKDIR}/genes"

gene=`ls ${GENEDIR} | sed "${SLURM_ARRAY_TASK_ID}q;d"`
mkdir ${GENEDIR}/${gene}/paml

module load miniconda
source activate /home/uvi/be/avs/tools/miniconda_avs

echo "Running Codeml for ${gene}"
ete3 evol -t ${GENEDIR}/${gene}/tree/${gene}_trimmed.raxml.bestTree --alg ${GENEDIR}/${gene}/align/${gene}_align_DNA_trimmed2.fasta 
--codeml_param CodonFreq,3 ncatG,4 verbose,0  --models M0 M3 M1 M2 M7 M8 M8a --cpu 7 -o ${GENEDIR}/${gene}/paml > ${GENEDIR}/${gene}/paml/${gene}_test_M7-M8.out
```
In this step, I discarded the genes *HAP90AA1, ERG* because they did not complete successfully all site-models.

```{r load_data, eval=F}
load("data/evoldata")
load("data/fundata")

idx <- sapply(as.character(evoldata[,1]),function (x) which(as.character(fundata[,1])==x))
fundata<-fundata[idx,]
fulldata<-cbind(fundata,evoldata[,-1])
```

##Global dN/dS distribution
```{r _global_dNdS_sist, eval=T,fig.align='center', dpi=100, fig.width=8, fig.height=6,fig.cap="Figure . Histogram showing distribution of global dN/dS estimates acroos 437 cancer genes"}
#Combine functional and evolutionary data frames

hist(fulldata$avomega,main=NULL,xlab="Global dN/dS",col="lightblue")
```
The genes with highest global dN/dS (> 0.5) estimates were:

* Interleukin 2 (IL-2): protein required for T-cell proliferation and other activities crucial to regulation of the immune response. Involved in intestinal T-cell lymphoma. global dN/dS = 0.748.
* Tumor necrosis factor receptor superfamily member 6 (FAS): a tumor supressor gene involved in apoptosis induction. global dN/dS = 0.601.
* Fc receptor-like protein 4 (FCRL4): may function in the B-cell-mediated immune response. Global dN/dS = 0.6.
* NUT family member 2A (NUTM2A): No funcitonal information. Global dN/dS = 0.573.
* Partner and localizer of BRCA2 (PALB2): plays a critical role in homologous recombination repair (HRR) through its ability to recruit BRCA2 and RAD51 to DNA breaks. Global dN/dS = 0.507.
* Programmed cell death 1 ligand 2 (PDCD1LG2): Involved in the costimulatory signal, essential for T-cell proliferation and IFNG production. Global dN/dS = 0.506.

#Correlation analysis
I correlate global dN/dS with the categories:

* Mutation type (germinal/somatic)
* Tissue type
* Mutation genetics (recesive/dominant)
* Cancer role
* Mutation impact

```{r load dndspercat_function,eval=T}
###dndspercat###
#Compare global dN/dS among functional categories embeded in a column
#Generate a concatenated data frame of all categories and a boxplot with significance values 
dndspercat<-function(category,names,df,column) {
  
  dfout<-data.frame()
  
  for (i in 1:length(category)) {
    tmp<-df[grep(category[i],column),c("gene","avomega","pos_selec")]
    tmp$class=rep(names[i])
    dfout<-rbind(dfout,tmp)
  }
  
  xsq<-chisq.test(table(dfout$class,dfout$pos_selec))
  
  #Boxplot comparing global dN/dS
  mat<-combn(names,2)
  comp<-lapply(seq_len(ncol(mat)), function(i) mat[,i])
  dndsboxpl<-ggboxplot(dfout, x = "class", y = "avomega", select= names, color="class",palette="jco",xlab=FALSE, ylab = "Global dN/dS") + stat_compare_means(comparisons = comp,label="p.signif")
  
  return(list(df=dfout,chisq=xsq,dndsplot=dndsboxpl))
}

```

```{r load_fulldata, eval =T}
load("data/fulldata")
```

```{r coor_global_dnds, eval=T}
#Statistical analysis
library(ggpubr)
attach(fulldata)

#Mutation class
mutclasses<-unique(unlist(strsplit(as.character(fulldata$mut.class),split=", ")))
mutclassnames<-c("Somatic","Som & Germ", "Germinal")
mutclassout<-dndspercat(mutclasses,mutclassnames,fulldata,mut.class)

#Tissue type
tissues<-unique(unlist(strsplit(levels(fulldata$tissue.type),split=",|;")))
tisnames<-c("Epithelial","Leuk/Lymp","Mesenchymal","Other")
tisout<-dndspercat(tissues,tisnames,fulldata,tissue.type)

#Genetics
genet<-levels(fulldata$genetics)
genet<-c("Dom","Rec")
genames<-c("Dominant","Recessive")
genout<-dndspercat(genet,genames,fulldata,genetics)

#Cancer role
canrole<-gsub("\"","",levels(fulldata$cancer.role))
canrole<-unique(unlist(strsplit(canrole,", ")))
canout<-dndspercat(canrole, canrole,fulldata,cancer.role)

#Mutation type (functional)
mutype<-gsub("\"","",fulldata$mutation.type)
mutype<-unique(unlist(strsplit(mutype,", |. |,| ")))
mutype2<-c("Mis","N")
mutnames<-c("Missense","Nosense")
mutout<-dndspercat(mutype,mutnames,fulldata,mutation.type)
```

```{r dnds_muts_plot, eval=T, fig.align="center", dpi=100, fig.width=10, fig.height=7, fig.cap="Figure . Boxplot showing global dN/dS for genes carrying or not germinal mutations. Right plot is faceted by presence or absence of somatic mutations in the genes"}
ger_plot<-ggboxplot(fulldata, x = "germline", y = "avomega",color="germline",palette="jco",ylab = "Global dN/dS") + stat_compare_means(method="wilcox.test",label.x=1.5,cex=5)
ger_plot2<-ggboxplot(fulldata, x = "germline", y = "avomega",color="germline",palette="jco",facet.by="somatic",ylab = "Global dN/dS", short.panel.labs = F) + stat_compare_means(method="wilcox.test",label.x=1.5,cex=4)

ggarrange(ger_plot,ger_plot2, ncol=2,nrow=1,widths=c(2,3))
```
We observed that those genes for which no somatic mutation but yes germline mutations have been identified (36 genes) show higher global dN/dS.

```{r nosom_yesgerm}
nosomyesgerm<-fundata[fundata$somatic == "no" & fundata$germline == "yes","gene"]
```

```{r dnds_tis_plot, eval=T, fig.align="center", dpi=100, fig.width=7, fig.height=7,fig.cap="Figure . Boxplot comparing global dN/dS for genes associated with different tissue types. Statistical significance for each pairwise comparison between categories is indicated"}
tisout$dndsplot
```

```{r dnds_genet_plot, eval=T, fig.align="center", dpi=100, fig.width=7, fig.height=7,fig.cap="Figure . Boxplot comparing global dN/dS for genes carrying mutations with different genetics. Statistical significance between categories is indicated"}
genout$dndsplot
```

```{r dnds_canrole_plot, eval=T, fig.align="center", dpi=100, fig.width=7, fig.height=7,fig.cap="Figure . Boxplot comparing global dN/dS for genes associated with different roles in cancer. Statistical significance of comparisons between categories is indicated"}
canout$dndsplot
```

```{r dnds_mutef_plot, eval=T, fig.align="center", dpi=100, fig.width=7, fig.height=7,fig.cap="Figure . Boxplot comparing global dN/dS for genes carrying muations with different funcitonal impact. Statistical significance of comparisons between categories is indicated"}
mutout$dndsplot
```

We also compared evolution of cancer genes between somatic and germinal levels. For this we took data from Martín-Corena et al. (2017) and tested wheter global dN/dS vary in genes under positive selection at the somatic level.

```{r ps_plot, eval=T, fig.align="center", dpi=100, fig.width=7, fig.height=14,fig.cap="Figure . Up: Boxplot comparing global dN/dS for genes previously identified (or not) under positive somatic selection. Statistical significance of comparison is indicated. Botton: correlation between germinal and somatic (missense) dN/dS. Blue points represent genes under soamtic positive selection according to MartínCorena et al. 2017." }
attach(fulldata)
psplot<-ggboxplot(fulldata, x = "driver_mc", y = "avomega",color="driver_mc",palette="jco",ylab = "Global dN/dS", short.panel.labs = F,xlab="Positive selection",submain="Driver genes from Martín-Corena et al. 2017") + stat_compare_means(method="wilcox.test",label.x=1.5,cex=5)

somgerplot<-ggplot(fulldata,aes(avomega,log10(dnds_mc),colour=driver_mc))+geom_point() + xlab("germinal dN/dS") +ylab("log10(somatic dN/dS)") + annotate(x=0.5, y=1.5, label=paste("R = ", round(cor(avomega, log10(dnds_mc)),2)),geom="text", size=5)

ggarrange(psplot,somgerplot,nrow=2,ncol=1)
```
##Selection test (site-models)
We identified 96 genes under positive selection (see **positive_selection_cancer_genes** table)

```{r positive_selection_genes,eval=T, comment=F, warning=F}
psgenes<-fulldata[fulldata$lrtM1M2 < 0.05 | fulldata$lrtM8M8a < 0.05,]
psgenes2<-psgenes[order(psgenes$lrtM8M8a,psgenes$lrtM1M2),c(1,2,5,20,25,26,27)]
library(knitr)
kable(psgenes2,row.names = F,caption = "Table. Cancer genes under positive selection")
```


Positively selected genes are overrepresented in DNA recombination and DNA repair (see *GO_overrepresentation_selected_genes.txt* file).
![Figure. GO enrichment for Biological proccesses of cancer genes under positive selection](GOanalysis/GOPROCESS.png)

Among these genes, we find BRCA2, Involved in double-strand break repair and/or homologous recombination and associated with several tumour types. Interestingly, among selected genes there are 5 members of the "Fanconi anemia complementation group (FANCD2, FANCG, FANCA,FANCE, FANCC). Eighty to 90 percent of cases of Fanconi anemia are due to mutations in one of three genes, FANCA, FANCC, and FANCG.

We compared the proportion of positively selected genes among functional categories
```{r compar_prop_psgenes, eval=T, fig.align="center", dpi=100, fig.width=10, fig.height=10,fig.cap="Figure . Comparison of proportion of genes under positive selection among categories. The significance value for a chi-square test is shown below"}

par(mfrow=c(2,2))
mosaicplot(table(fulldata$mut.class,fulldata$pos_selec),ylab="Positive selection",color = c("lightblue","red"),main = "Mutation class",cex.axis =1,xlab=paste("p-value(chisq.test)",round(mutclassout$chisq$p.value,3),sep="="))

mosaicplot(table(tisout$df$class,tisout$df$pos_selec),ylab="Positive selection",color = c("lightblue","red"),main = "Tissue type",cex.axis =1,xlab=paste("p-value(chisq.test)",round(tisout$chisq$p.value,3),sep="="))

mosaicplot(table(mutout$df$class,mutout$df$pos_selec),ylab="Positive selection",color = c("lightblue","red"),main = "Mutation type",cex.axis =1,xlab=paste("p-value(chisq.test)",round(mutout$chisq$p.value,3),sep="="))

mosaicplot(table(genout$df$class,genout$df$pos_selec),ylab="Positive selection",color = c("lightblue","red"),main = "Genetics",cex.axis =1,xlab=paste("p-value(chisq.test)",round(genout$chisq$p.value,3),sep="="))
```

Out of these 96 genes, 20 (21%) overlap with those genes with exclusively germline mutations. This seems significantly higher than the expected proportion of 8.2% for the presence of these genes.
```{r overlap_sel_germ_genes}
psgenes2[psgenes2$gene %in% nosomyesgerm, 1]
```
Within this group are 4 FANC members (FANCD2, FANCG, FANCA, FANCE).

#Testing positive selection in human branch
324 alignments which contain human ortholog were tested for episodic selection in human branch applying **branch-site models**
```{bash bs_human}
#!/bin/bash

#SBATCH -N 1
#SBATCH -n 1
#SBATCH -c 2
#SBATCH --cpus-per-task 2
#SBATCH -t 05:00:00
#SBATCH --mem 6G

#Setting directories and files
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
GENEDIR="${WORKDIR}/genes"

gene=`ls ${GENEDIR} | sed "${SLURM_ARRAY_TASK_ID}q;d"`

HUMID=`grep -o 'ENSG[0-9]*' ${GENEDIR}/${gene}/tree/${gene}_trimmed.raxml.bestTree` > tmp
echo "the human id for $gene is $HUMID"

module load miniconda
source activate /home/uvi/be/avs/lustre/tools/miniconda_avs

echo "Running Codeml for ${gene}"
ete3 evol -t ${GENEDIR}/${gene}/tree/${gene}_trimmed.raxml.bestTree --alg ${GENEDIR}/${gene}/align/${gene}_align_DNA_trimmed2.fasta --codeml_param CodonFreq,3 ncatG,4 verbose,0 --models bsA bsA1 
 --mark $HUMID --cpu 2 -o ${GENEDIR}/${gene}/paml > ${GENEDIR}/${gene}/paml/${gene}_bs_human.out
```
We did not detect evidence of positive selection on human for any gene.