---
title: "Selective pressures on cancer genes in mammals"
author: "Alberto Vicens Sanchez"
date: "January 24, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, eval = FALSE, warnings = FALSE, cache = TRUE, cache.path="/home/user/phylogenomics_lab/selection_cancer_genes/cache/")
```

Although many multicellular organisms share common cancer resistance pathways, different species have evolved independent mechanism of tumor supression. It is therefore important to understand how selection has supressed cancer in different evolutionary stages to ultimately improve cancer prevention in humans. 

In this study, we will evaluate the selective pressures driving the evolution of human cancer associated genes, spanning both oncogenes and tumor supressor genes (TSG), in a phylogeny of mammals. 

The objectives of this study are:

* Assess selective pressures on cancer driver genes in mammals.
* Compare selective regimes among long-lived and short-lived animals.
* Compare selective regimes between oncogenes and TSG.
* Compare adaptive germline variation against somatic mutations.

#1. Data collection
## 1.1) Cancer genes
For this study, I analyze **102 cancer genes** cancer genes with reported cases of germline mutations. We retrieved these genes from COSMIC database after filtering for germline mutations (http://cancer.sanger.ac.uk/cosmic/census/tables?name=gmuts).

A table with cancer genes and detailed information can be viewed in this link:
https://docs.google.com/spreadsheets/d/1UmkyjbfW1ju0qnO4uZJw0hUMljUczLJuipcIU_byTFY/edit?usp=sharing

We retrieve the Entrez ID's column from the COSMIC table and convert it to RefSeq protein ID using DAVID Gene ID Conversion tool(https://david.ncifcrf.gov/conversion.jsp)

We obtained 642 protein IDs due to the multiple isoforms annotated for genes. To retrieve a single protein sequence per gene, we filtered out repeated protein annotations
```{bash}
cut -f2,4 david_conv_refseq_genid_protid.tsv | tail -n +2 | sort -k2 | uniq -2 |cut -f1 > proteinid.txt
```
To retrieve sequences from the protein ID list, we write the perl script *retrieve_sequences.pl*.

```{perl retrieve_sequences}
#!usr/bin/perl

use strict;
use warnings;
use Bio::Perl;

my $filename = 'proteinid.txt';

unless (open(PROTEINID, $filename)) {
	print "Could not open the file $filename !\n";
	exit;
}

my @protids = <PROTEINID>;
close PROTEINID;

my $protname = ' ' ;
my $seq_object = ' ';

foreach my $prot (@protids) {
	chomp $prot;
	$protname = join "", ">",$prot,".pep";
	$seq_object = get_sequence('refseq', $prot);
	write_sequence($protname, 'fasta', $seq_object);
	
}

exit;
```
Next, we create a folder for each gene with the script *create_gene_folders.sh* 

```{bash create_gene_folders}
#!/bin/bash

DATADIR="/home/uvi/be/avs/cancer_genes_selection/datafiles/"
GENEDIR="/home/uvi/be/avs/cancer_genes_selection/datagenes/"
PROTEINLIST="${DATADIR}/proteinid.txt"

while read id; do
mkdir ${DATADIR}/`grep "$id" ${GENEDIR}/david_conv_refseq_genid_protid.tsv | cut -d"(" -f2 | sed 's/)$//'`
done < ${PROTEINLIST}
```

We then move the retrieved human protein sequences to the respective gene folder with the script *move_sequences_to_folders.sh*.
```{bash move_sequences_to_folders}
#!/bin/bash

SEQDIR="/home/uvi/be/avs/cancer_genes_selection/dataseqs/"
GENEDIR="/home/uvi/be/avs/cancer_genes_selection/datagenes/"
PROTEINLIST="${GENEDIR}/proteinid.txt"

while read id; do
GENENAME=`grep "$id" ${GENEDIR}/david_conv_refseq_genid_protid.tsv | cut -d"(" -f2 | sed 's/)$//'`
PROTFILE=`ls ${SEQDIR} | grep "$id"`
mv ${SEQDIR}/${PROTFILE} ${SEQDIR}/${GENENAME}/
done < ${PROTEINLIST}
```

## 1.2) Mammalian databases
We chose 67 mammalian genomes plus 2 vertebrate (*Gallus gallus* and *Anolis carolinensis*) as outgroup for the study.

A table with the genomes used for this study and detailed information can be viewed in:
https://docs.google.com/spreadsheets/d/1b9wOk1Hao8mVTIaU_DkpyuhoA5eekG6BudykRStExog/edit?usp=sharing

We download the proteomes from FTP site of Ensembl (ftp://ftp.ensembl.org/) with the script *download_genomes.sh*

```{bash download_genomes}
#!usr/bin/bash
USER="anonymous"
PASSWD="password=chechu086@gmail.com"
HOST="ftp://ftp.ensembl.org/pub/release-91/fasta/"
SPLIST="/home/uvi/be/avs/cancer_genes_selection/species_name.txt"
DBDIR="/home/uvi/be/avs/cancer_genes_selection/databases/pep"

while read name; do

wget --user=${USER} --password=${PASSWD} "${HOST}/$(echo "$name" | awk '{print tolower($0)}')/pep/*.gz" -P ${DBDIR}
done < ${SPLIST}
```

*NOTE:We then uncompress the genome files and edit the sequence headers, substituting spaces by pipes.*
```{bash}
cd databases/cds/ | gunzip -v *.gz
for db in *.fa; do sed -i.bak 's/ /|/g' $db; done
```

*NOTE: Move the backup files (.bak) to $LUSTRE*
```{bash}
mkdir $LUSTRE/cancer_genes_selection/databases/cds/
mv $HOME/cancer_genes_selection/databases/cds/*.bak \ $LUSTRE/cancer_genes_selection/databases/cds/
```

#2. Getting othologues
## 2.1) Making BLAST databases
We build BLAST databases for genomic files from the studied species runnung the *make_blast_database.sh* script

```{bash make_blast_db}
#!/usr/bin/bash

#Set initial directories and files
DBDIR="/home/uvi/be/avs/cancer_genes_selection/databases/pep"
DBFASTA=`ls ${DBDIR} | grep '.fa$'`

#Load modules
module load gcc/5.3.0 ncbi-blast/2.2.31+

while read -r fasta; do
makeblastdb -in "${DBDIR}/$fasta" -dbtype prot -out "${DBDIR}/$fasta"
done <<< "$DBFASTA"
```

## 2.2)Reciprocal BLASTp
With the genomic databases saved in the folder *databases* and human protein sequences saved in per-gene subfolders, within the higher folder *datagene*, we perform **tblastn** for each gene against all species databases. For this, we  *run_tblastn_cesga.sh* script.

```{bash run_blastp}
#!/usr/bin/bash

GENENAMES=`ls ${GENEDIR}`

while read -r gene; do

echo "Running blastp for $gene"
mkdir ${GENEDIR}/${gene}/blastp

        for db in ${DBDIR}/*.fa; do
        DBNAME=`echo ${db##*/} <<< ${db}`
        SPNAME=`cut -d"." -f1 <<< ${DBNAME}`
        echo "Exploring database ${DBNAME}"
        blastp -query ${GENEDIR}/${gene}/*.pep -db ${db} -outfmt '6 qseqid sseq$
done
done <<< "$GENENAMES"
```

We then extract the first hits of each BLAST search with the *extract_firsthits.sh* script.

```{bash extract_firsthits}
#!/usr/bin/bash

GENEDIR="/home/uvi/be/avs/cancer_genes_selection/datagenes"
GENENAMES=`ls ${GENEDIR}`

while read -r gene; do
echo "Extracting potential orthologs for $gene"
mkdir ${GENEDIR}/${gene}/recblast
mkdir ${GENEDIR}/${gene}/recblast/firsthits
SEQDIR="${GENEDIR}/${gene}/recblast/firsthits"

for file in ${GENEDIR}/${gene}/blastp/*.tsv; do
SEQNAME=`echo ${file##*/} | sed 's/blastout.tsv/firsthit.fas/'`
head -1 $file | cut -f2,7 | sed 's/-//g' | awk 'BEGIN{FS="\t";} {print ">"$1"\n"$2;}' > ${SEQDIR}/${SEQNAME}
done
done <<< "$GENENAMES"
```

We then perform BLASTp searches of best hits against human protein database. 
Before BLAST, we modified the human db to get the full annotation of sequences. In this way, we will able filter the results by the gene symbol of the studied gene.
```{bash edit_human_db}
cd /home/uvi/be/avs/cancer_genes_selection/databases/
< pep/Homo_sapiens.GRCh38.pep.all.fa tr ' ' '_' > Homo_sapiens.GRCh38.pep2.all.fa
makeblastdb -in Homo_sapiens.GRCh38.pep2.all.fa -dbtype prot -parse_seqids
```

With list of best hit protein files and the edited human database, we run the second BLAST with the script **reciprocal_blast.sh**
```{bash rec_blast}
#!/usr/bin/env bash
GENEDIR="/home/user/phylogenomics_lab/cancer_genes_selection/datagenes"
DBDIR="/home/user/phylogenomics_lab/cancer_genes_selection/databases"
GENENAMES=`ls ${GENEDIR}`

while read -r gene; do
SEQDIR="${GENEDIR}/${gene}/recblast/firsthits"
HITFILE="${gene}_firsthits.fas"
echo "Concatenating best hits from each species"
cat ${SEQDIR}/*.fas > ${GENEDIR}/${gene}/recblast/${HITFILE}
echo "Running Blast for ${gene}"
blastp -query ${GENEDIR}/${gene}/recblast/${HITFILE} -db ${DBDIR}/Homo_sapiens.GRCh38.pep2.all.fa -outfmt 6 -max_target_seqs 1 > ${GENEDIR}/${gene}/recblast/"${gene}_recblast_first_hits.tsv"

done <<< "$GENENAMES"
```

## 2.3) Downloading sequences from protein id. 
We extract the IDs of ortholog proteins (those matching with the target human gene).
```{bash retrieve_ortholog_ids}
#!/usr/bin/bash

GENEDIR="/home/user/phylogenomics_lab/cancer_genes_selection/datagenes"
GENENAMES=`ls ${GENEDIR}`

while read -r gene; do

BLASTOUT="${GENEDIR}/${gene}/recblast/${gene}_recblast_first_hits.tsv"
ORTHOLIST=${gene}_orthologs.txt

cat ${BLASTOUT} | grep ${gene} | cut -f1 | uniq > ${GENEDIR}/${gene}/$ORTHOLIST
```

Next, I download coding and protein ortholog sequences. For this we run the script ****
```{bash download_cds-pep_from_protid}
#!/bin/bash

WORKDIR="/home/user/phylogenomics_lab/cancer_genes_selection"
DBDIR="/home/user/phylogenomics_lab/cancer_genes_selection/databases"
GENEDIR="/home/user/phylogenomics_lab/cancer_genes_selection/datagenes"
SCRIPTDIR="/home/user/phylogenomics_lab/cancer_genes_selection/scripts"

species="${WORKDIR}/species_codes.txt"
GENENAMES=`ls ${GENEDIR}`

while read -r gene; do
echo "Creating directory for coding sequences of ${gene}"
mkdir ${GENEDIR}/${gene}/seqs
mkdir ${GENEDIR}/${gene}/seqs/cds

protlist="${GENEDIR}/${gene}/*_orthologs.txt"

while read -r id; do
spcode=`echo $id | cut -c -6`
dataset=`grep $spcode ../species_codes.txt | cut -d' ' -f2`
echo "Downloading ortholog coding sequences of ${gene} from ${dataset}"
perl ${SCRIPTDIR}/download_cds_from_protid.pl $id $dataset >> ${GENEDIR}/${gene}/seqs/cds/"${gene}_cds.fas"
echo "Downloading ortholog peptide sequences of ${gene} from ${dataset}"
perl ${SCRIPTDIR}/download_pep_from_protid.pl $id $dataset >> ${GENEDIR}/${gene}/seqs/pep/"${gene}_pep.fas"
done < ${protlist}
echo ${gene}
done <<< "${GENENAMES}"
```

Within this bash script we call the perl script **download_cds_from_protid.pl** and **download_pep_from_protid.pl**, which download coding and peptide sequences for a given protein id.

#3) Multiple alignment

##3.1) Refinement pre-alignment
Once I have retrieved multiple sequence files, I remove the ambiguity sites (N) of sequences
```{bash remove_ambiguity_sites}
#!/usr/bin/bash

GENEDIR="/home/user/phylogenomics_lab/cancer_genes_selection/datagenes"
GENENAMES=`ls ${GENEDIR}`

while read -r gene; do

echo "Removing ambiguity sites for $gene"
sed '/^[^>]/s/N//g'  ${GENEDIR}/${gene}/seqs/cds/"${gene}_cds.fas" > ${GENEDIR}/${gene}/seqs/cds/"${gene}_cds_refined.fas"

done <<< "${GENENAMES}"
```

##3.2)Sequence alignment


