---
title: "Branch-site analysis of Cancer Genes"
author: "Alberto Vicens"
date: "06/22/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, eval = F, warning = F, comment = F)
```

To indentify cancer genes under positive selection in specific lineages, I conducted *Branch-Site tests* of positive selection. 

## Running Branch-Site models
Branch-site tests compare neutral and selection site-models on each branch in the phylogeny. In this case, we just analyzed the terminal branches leading to each lineage (internal branches were not evaluated). 
I run the analysis using ETE3 package (Huerta-Cepas et al. 2016) and jobs were submitted to the *CESGA* computing cluster.

```{bash, bs_test}
#!bin/bash
                              
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
GENEDIR="${WORKDIR}/genes"

gene=`ls ${GENEDIR} | sed "${SLURM_ARRAY_TASK_ID}q;d"`

module load miniconda
source activate /home/uvi/be/avs/lustre/tools/miniconda_avs

echo "Running Codeml for ${gene}"
$TMPDIR} --resume > ${GENEDIR}/${gene}/paml/${gene}_bs_test.out
```

## Data processing

### Getting nodes ID
In this step, I extract the node IDS assigned for each terminal branch in the Branch-Site analysis.

First, I get a table of correspondence between node IDS and species name with the *--node_ids* option of ETE3.
```{bash, get_node_ids}
#!/bin/bash

#Setting directories and files
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
GENEDIR="${WORKDIR}/genes"
GENENAMES=`ls ${GENEDIR}`

module load miniconda
source activate /home/uvi/be/avs/lustre/tools/miniconda_avs

while read -r  gene; do

echo "Getting node IDs for ${gene}"
ete3 evol -t ${GENEDIR}/${gene}/tree/${gene}_trimmed.raxml.bestTree --alg ${GENEDIR}/${gene}/align/${gene}_align_DNA_trimmed2.fasta --node_ids > ${GENEDIR}/${gene}/paml/${gene}_node_ids.txt
done <<< "${GENENAMES}"
```

I parse the output files to get only the table of correspondence. For this, I call iteratively to the perl script **extract_nodeIDs.pl**
```{bash, extract_nodeIDs}
#!/bin/env bash

#Setting directories and files
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
GENEDIR="${WORKDIR}/genes"
SCRIPTDIR="${WORKDIR}/scripts"
GENENAMES=`ls ${GENEDIR}`

while read -r gene; do
echo "Extracting node IDs for ${gene}"
BSPATH=${GENEDIR}/${gene}/paml/${gene}_node_ids.txt
perl extract_nodeIDs.pl $BSPATH > tmp
sed -e 's/~.\{33\}//g'  -e 's/\s//g' -e '2d' tmp > ${GENEDIR}/${gene}/paml/${gene}_nodeIDs_parsed.txt
rm tmp
done <<< "${GENENAMES}"
```
### Getting LRTs
Like I did for node IDs, I extracted the table with the LRT results from the output files of Branch-site tests calling the perl script **extract_LRT.pl**

```{bash, extract_LRTs}
#!/bin/env bash

#Setting directories and files
WORKDIR="/mnt/lustre/scratch/home/uvi/be/avs/cancer_genes_selection"
GENEDIR="${WORKDIR}/genes"
SCRIPTDIR="${WORKDIR}/scripts"
GENENAMES=`ls ${GENEDIR}`

while read -r gene; do
echo "Extracting LRT of Branch-site models for ${gene}"
BSPATH=${GENEDIR}/${gene}/paml/${gene}_bs_test.out
perl extract_LRT.pl $BSPATH > tmp
sed -e 's/~.\{33\}//g'  -e 's/\s//g' -e '2d' tmp > ${GENEDIR}/${gene}/paml/${gene}_LRT.txt
rm tmp
done <<< "${GENENAMES}"

```

